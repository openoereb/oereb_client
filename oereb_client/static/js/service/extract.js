goog.provide('oereb.ExtractService');

goog.require('oereb');

/**
 * Angular service to request a specific extract.
 * @param {angular.$http} $http Angular service for HTTP requests.
 * @param {angular.$q} $q Angular service for deferrable objects.
 * @param {angular.Module.constant} oerebApplicationUrl Angular service for HTTP requests.
 * @constructor
 * @ngInject
 * @ngdoc service
 * @ngname ExtractService
 */
oereb.ExtractService = function($http, $q, oerebApplicationUrl) {

  this.$http_ = $http;
  this.$q_ = $q;
  this.applicationUrl_ = oerebApplicationUrl;

  this.extract_ = undefined;
  this.embeddable_ = undefined;

};

/**
 * Requests the extract for the specified EGRID.
 * @param {string} egrid The EGRID to request the extract for.
 * @returns {angular.$q.Promise} Promise for the EGRID request.
 */
oereb.ExtractService.prototype.queryExtractById = function(egrid) {
  this.extract_ = undefined;
  this.embeddable_ = undefined;
  var def = this.$q_.defer();
  this.$http_.get(this.applicationUrl_ + 'extract/reduced/json/geometry/' + egrid).then(
    function(response) {
      if (angular.isObject(response.data['GetExtractByIdResponse'])) {
        this.extract_ = response.data['GetExtractByIdResponse']['extract'];
        if (angular.isObject(response.data['GetExtractByIdResponse']['embeddable'])) {
          this.embeddable_ = response.data['GetExtractByIdResponse']['embeddable'];
        }
        def.resolve(response.data['GetExtractByIdResponse']);
      }
      else {
        def.reject('Invalid response format.');
      }
    }.bind(this),
    function(response) {
      var error = 'Requesting extract for ' + egrid + ' failed.';
      if (angular.isString(response.data)) {
        error += ' ' + response.data;
      }
      def.reject(error);
    }
  );
  return def.promise;
};

/**
 * Returns the extract if available.
 * @returns {Object|undefined} The extract object or undefined.
 */
oereb.ExtractService.prototype.getExtract = function() {
  return this.extract_;
};

/**
 * Returns the real estate data if available.
 * @returns {Object|undefined} The real estate data or undefined.
 */
oereb.ExtractService.prototype.getRealEstate = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['RealEstate'];
  }
  return undefined;
};

/**
 * Returns list of concerned themes.
 * @returns {Array|undefined} The list of concerned themes.
 */
oereb.ExtractService.prototype.getConcernedThemes = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['ConcernedTheme'];
  }
  return undefined;
};

/**
 * Returns list of not concerned themes.
 * @returns {Array|undefined} The list of not concerned themes.
 */
oereb.ExtractService.prototype.getNotConcernedThemes = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['NotConcernedTheme'];
  }
  return undefined;
};

/**
 * Returns list of themes without data.
 * @returns {Array|undefined} The list of themes without data.
 */
oereb.ExtractService.prototype.getThemesWithoutData = function() {
  if (angular.isDefined(this.getExtract())) {
    return this.getExtract()['ThemeWithoutData'];
  }
  return undefined;
};

/**
 * Returns the public law restrictions of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Array|undefined} The public law restrictions of the specified topic.
 */
oereb.ExtractService.prototype.getRestrictions = function(themeCode) {
  if (angular.isDefined(this.getRealEstate())) {
    var restrictions = [];
    angular.forEach(this.getRealEstate()['RestrictionOnLandownership'], function(restriction) {
      if (restriction['Theme']['Code'] === themeCode) {
        restrictions.push(restriction);
      }
    });
    return restrictions;
  }
  return undefined;
};

/**
 * Returns the legend entries of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Object|undefined} The legend entries of the specified topic.
 */
oereb.ExtractService.prototype.getLegend = function(themeCode) {
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isDefined(restrictions)) {
    var legendEntries = [];
    var legendGraphics = [];
    for (var i = 0; i < restrictions.length; i++) {
      var existing = false;
      for (var j = 0; j < legendEntries.length; j++) {
        if (legendEntries[j]['TypeCode'] === restrictions[i]['TypeCode']) {
          existing = true;
          legendEntries[j]['Area'] += restrictions[i]['Area'];
          legendEntries[j]['PartInPercent'] += restrictions[i]['PartInPercent'];
          break;

        }

      }
      if (!existing) {
        legendEntries.push({
          'TypeCode': restrictions[i]['TypeCode'],
          'Information': restrictions[i]['Information'],
          'Area': restrictions[i]['Area'],
          'PartInPercent': restrictions[i]['PartInPercent'],
          'SymbolRef': restrictions[i]['SymbolRef']
        });
      }
      this.addIfNotContains_(restrictions[i]['Map']['LegendAtWeb'], legendGraphics);
    }
    return {
      'entries': legendEntries,
      'graphics': legendGraphics
    };
  }
  return undefined;
};

/**
 * Adds the item to the specified target array if it's not already contained. `angular.equals` is used for the
 * comparison.
 * @param {Object|Array|string|number} item The element to be added.
 * @param {Array} target The target array to add the element to.
 * @private
 */
oereb.ExtractService.prototype.addIfNotContains_ = function(item, target) {
  if (angular.isArray(target)) {
    for (var i = 0; i < target.length; i++) {
      if (angular.equals(target[i], item)) {
        return;
      }
    }
    target.push(item);
  }
};

/**
 * Adds the document to the specified target array if it's not already contained. `angular.equals` is used for
 * the comparison.
 * @param {Object} document The document to be added.
 * @param {Array} target The target array to add the element to.
 * @private
 */
oereb.ExtractService.prototype.addDocumentIfNotContained_ = function(document, target) {
  for (var i = 0; i < target.length; i++) {
    // Check if document already exists (only use 'Title', 'OfficialNumber' and 'TextAtWeb' for comparison)
    if (
      angular.equals(target[i]['Title'], document['Title']) &&
      angular.equals(target[i]['OfficialNumber'], document['OfficialNumber']) &&
      angular.equals(target[i]['TextAtWeb'], document['TextAtWeb'])
    ) {
      // Add missing article numbers if document already exists
      for (var j = 0; j < document['ArticleNumber'].length; j++) {
        this.addIfNotContains_(document['ArticleNumber'][j], target[i]['ArticleNumber']);
      }
      // Add missing articles if document already exists
      for (var k = 0; k < document['Article'].length; k++) {
        this.addIfNotContains_(document['Article'][k], target[i]['Article']);
      }
      return;
    }
  }
  // Add document if it is missing
  target.push(document);
};

/**
 * Add each listed document to the target array if it's not already contained.
 * @param {Array} documents The documents to be added.
 * @param {Array} target The target array.
 * @private
 */
oereb.ExtractService.prototype.addDocumentsIfNotContained_ = function(documents, target) {
  // Iterate documents to be added
  for (var i = 0; i < documents.length; i++) {
    // Create document object with necessary properties and check if it needs to be added
    this.addDocumentIfNotContained_({
      'Title': documents[i]['Title'],
      'OfficialNumber': documents[i]['OfficialNumber'],
      'ArticleNumber': documents[i]['ArticleNumber'] || [],
      'Article': documents[i]['Article'] || [],
      'TextAtWeb': documents[i]['TextAtWeb']
    }, target);
    // Do the same for possible references
    var references = documents[i]['Reference'];
    if (angular.isArray(references)) {
      this.addDocumentsIfNotContained_(references, target);
    }
  }
};

/**
 * Returns the legal documents of the specified topic.
 * @param {string} themeCode The code of the topic to return the restrictions for.
 * @returns {Object|undefined} The legal documents of the specified topic.
 */
oereb.ExtractService.prototype.getDocuments = function(themeCode) {
  // Get the restrictions for the specified topic
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isArray(restrictions)) {
    var legalProvisions = [];
    var documents = [];
    // Iterate the resulting restrictions
    for (var i = 0; i < restrictions.length; i++) {
      // Iterate the legal provisions for each restriction
      var legalProvision = restrictions[i]['LegalProvisions'];
      for (var j = 0; j < legalProvision.length; j++) {
        // Create legal provision object with necessary properties and check if it needs to be added
        this.addDocumentIfNotContained_({
          'Title': legalProvision[j]['Title'],
          'OfficialNumber': legalProvision[j]['OfficialNumber'],
          'ArticleNumber': legalProvision[j]['ArticleNumber'] || [],
          'Article': legalProvision[j]['Article'] || [],
          'TextAtWeb': legalProvision[j]['TextAtWeb']
        }, legalProvisions);
        // Add possible references to the documents
        var references = legalProvision[j]['Reference'];
        if (angular.isArray(references)) {
          this.addDocumentsIfNotContained_(references, documents);
        }
      }
    }
    // Add possible references for the real estate to the documents
    this.addDocumentsIfNotContained_(this.getRealEstate()['Reference'], documents);
    return {
      'LegalProvisions': legalProvisions,
      'Documents': documents
    };
  }
  return undefined;
};

/**
 * Returns the responsible offices of the specified topic.
 * @param {string} themeCode The code of the topic to return the responsible offices for.
 * @returns {Array|undefined} The responsible offices of the specified topic.
 */
oereb.ExtractService.prototype.getResponsibleOffices = function(themeCode) {
  var restrictions = this.getRestrictions(themeCode);
  if (angular.isArray(restrictions)) {
    var offices = [];
    for (var i = 0; i < restrictions.length; i++) {
      this.addIfNotContains_(restrictions[i]['ResponsibleOffice'], offices);
    }
    return offices;
  }
  return undefined;
};

/**
 * Creates a view service definition using the restriction's theme code and view service URL.
 * @param {string} themeCode The restriction's theme code.
 * @param {string} viewServiceUrl The restriction's view service URL.
 * @returns {Object} The created view service definition.
 * @private
 */
oereb.ExtractService.prototype.getViewServiceFromUrl_ = function(themeCode, viewServiceUrl) {
  var parts = viewServiceUrl.split('?');
  var url = parts[0];
  var definition = {
    'topic': themeCode,
    'url': url,
    'params': {}
  };
  var params = parts[1].split('&');
  for (var i = 0; i < params.length; i++) {
    var param = params[i].split('=');
    if (param[0].toUpperCase() === 'LAYERS') {
      definition['params']['LAYERS'] = param[1];
    }
    else if (param[0].toUpperCase() === 'STYLES') {
      definition['params']['STYLES'] = param[1];
    }
    else if (param[0].toUpperCase() === 'VERSION') {
      definition['params']['VERSION'] = param[1];
    }
  }
  return definition;
};

/**
 * Returns a unique list of view services available in the extract.
 * @returns {Array} A unique list of view service objects.
 */
oereb.ExtractService.prototype.getViewServices = function() {
  var viewServices = [];
  var realEstate = this.getRealEstate();
  if (angular.isDefined(realEstate)) {
    var restrictions = realEstate['RestrictionOnLandownership'];
    if (angular.isArray(restrictions)) {
      for (var i = 0; i < restrictions.length; i++) {
        var url = restrictions[i]['Map']['ReferenceWMS'];
        if (angular.isDefined(url)) {
          this.addIfNotContains_(
            this.getViewServiceFromUrl_(restrictions[i]['Theme']['Code'], url),
            viewServices
          );
        }
      }
    }
  }
  return viewServices;
};

/**
 * Returns the exclusions of liability for the current extract.
 * @returns {Array} The exclusions of liability.
 */
oereb.ExtractService.prototype.getExclusionsOfLiability = function() {
  if (angular.isDefined(this.getExtract()) && angular.isArray(this.getExtract()['ExclusionOfLiability'])) {
    return this.getExtract()['ExclusionOfLiability'];
  }
  return [];
};

/**
 * Returns the glossary for the current extract.
 * @returns {Array} The exclusions of liability.
 */
oereb.ExtractService.prototype.getGlossary = function() {
  if (angular.isDefined(this.getExtract()) && angular.isArray(this.getExtract()['Glossary'])) {
    return this.getExtract()['Glossary'];
  }
  return [];
};

/**
 * Returns the embeddable if available.
 * @returns {Object|undefined} The extract object or undefined.
 */
oereb.ExtractService.prototype.getEmbeddable = function() {
  return this.embeddable_;
};

oereb.module.service('ExtractService', oereb.ExtractService);
