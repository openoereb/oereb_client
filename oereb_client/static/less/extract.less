/**
 * Extract styling
 */

@import "variables";

div#extract {
  width: @menu-with;
  max-width: 100%;
  position: fixed;
  top: 0;
  left: 0;
  height: 0;
  padding-top: @header-height;
  background-color: #fff;
  transition: height @animation-duration;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  div.extract-wrapper {
    z-index: 1;
    background-color: @menu-background-color;
    overflow-y: scroll;
    flex: 1;
    padding-bottom: @default-padding;
    margin-top: @default-padding;
  }
  span.extract-close {
    z-index: 2;
    display: block;
    float: right;
    color: @brand-primary;
    opacity: 1;
  }
  span.extract-close:hover {
    color: #000000;
    opacity: 1;
  }
  .panel-default > .panel-heading .badge {
    background-color: @brand-primary;
  }
  .collapse-extract {
    padding-top: @default-padding;
    .btn {
      border-bottom-width: 0;
    }
  }
}

div#extract.loading {
  height: @loading-indicator-height;
}

div#extract.active {
  height: 100%;
}

div#extract.information {
  box-shadow: @menu-shadow;
}