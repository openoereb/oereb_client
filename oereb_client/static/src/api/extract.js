export function queryExtractById(applicationUrl, egrid) {
    const url = new URL(applicationUrl + 'extract/reduced/json/geometry/' + egrid);
    url.searchParams.append('_dc', new Date().getTime());
    return fetch(url).then(response => response.json());
}

export function sanitizeTopicCode(code) {
    return code.replace('.', '').replace('-', '').replace('_', '');
}

export function groupRestrictionsByTopic(retstrictions) {
    const restrictionsByTopic = {};
    retstrictions.forEach((restriction) => {
        const code = sanitizeTopicCode(restriction.Theme.Code);
        if (!restrictionsByTopic.hasOwnProperty(code)) {
            restrictionsByTopic[code] = [];
        }
        restrictionsByTopic[code].push(restriction);
    });
    return restrictionsByTopic;
}
