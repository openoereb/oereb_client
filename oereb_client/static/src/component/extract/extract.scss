@import "../../oereb_client.scss";

div.oereb-client-extract-wrapper {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    background-color: $menu-background-color;
    box-shadow: $default-shadow;
    overflow: hidden;
    transition: height $animation-duration linear;

    div.oereb-client-extract {
        width: 100%;
        height: 100%;
        padding-top: $header-height;
        overflow: hidden;
    }
    
    div.oereb-client-extract.data {
        .category:not(.show) {
            border: 0px;
        }
    
        & > div.accordion {
            overflow-y: scroll;
    
            button.accordion-button:focus,
            button.accordion-button:not(.collapsed) {
                box-shadow: none;
            }
        }

        @include media-breakpoint-up(xl) {
            button.collapse-extract {
                display: none !important;
            }
        }
    }
}

div.oereb-client-extract-wrapper.hidden {
    height: $header-height;
}

div.oereb-client-extract-wrapper.loading {
    height: $header-height + 100px;
}

div.oereb-client-extract-wrapper.shown {
    height: 100%;
}

div.oereb-client-extract-wrapper.collapsed {
    height: $header-height + 50px;
}

@include media-breakpoint-up(xl) {
    div.oereb-client-extract-wrapper {
        width: $menu-with;
    }
}