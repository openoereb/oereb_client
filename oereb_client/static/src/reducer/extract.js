import { createSlice } from "@reduxjs/toolkit";

export const extractSlice = createSlice({
    name: 'extract',
    initialState: {
        loading: false,
        visible: false,
        error: false,
        egrid: null,
        data: {}
    },
    reducers: {
        loadExtract: (state, action) => {
            console.log('Extract loading');
            if (!state.loading) {
                state.loading = true;
                state.visible = false;
                state.error = false;
                state.egrid = action.payload.egrid;
                state.data = {};
            }
        },
        showExtract: (state, action) => {
            state.loading = false;
            state.visible = true;
            state.error = false;
            state.data = action.payload.extract;
            console.log('Extract loaded', state.data);
        },
        showError: (state, action) => {
            state.loading = false;
            state.visible = false;
            state.error = true;
            state.data = {};
            console.log('Extract failed');
        },
        hideExtract: (state, action) => {
            state.loading = false;
            state.visible = false;
            state.error = false;
            state.egrid = null;
            state.data = {};
            console.log('Extract closed');
        }
    }
});

export const { loadExtract, showExtract, showError, hideExtract } = extractSlice.actions;

export default extractSlice.reducer;
