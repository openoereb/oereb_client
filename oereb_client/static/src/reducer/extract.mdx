import {createSlice} from "@reduxjs/toolkit";

export const extractSlice = createSlice({
  name: 'extract',
  initialState: {
    loading: false,
    zoom: false,
    visible: false,
    collapsed: false,
    error: false,
    egrid: null,
    identdn: null,
    number: null,
    information: false,
    tab: 0,
    data: {},
    highlight: true,
    background: true
  },
  reducers: {
    loadExtract: (state, action) => {
      if (!state.loading) {
        state.loading = true;
        state.visible = false;
        state.collapsed = false;
        state.error = false;
        state.information = false;
        state.tab = 0;
        state.egrid = action.payload.egrid;
        state.identdn = action.payload.identdn;
        state.number = action.payload.number;
        state.zoom = action.payload.zoom;
        state.data = {};
        const query = new URLSearchParams(window.location.search);
        if (state.egrid !== null) {
          query.set('egrid', state.egrid);
        }
        else if (state.identdn !== null && state.number !== null) {
          query.set('identdn', state.identdn);
          query.set('number', state.number);
        }
        window.history.pushState(null, null, '?' + query.toString());
      }
    },
    showExtract: (state, action) => {
      state.loading = false;
      state.visible = true;
      state.collapsed = false;
      state.error = false;
      state.information = false;
      state.highlight = true;
      state.background = true;
      state.tab = 0;
      state.data = action.payload;
    },
    showError: (state) => {
      state.loading = false;
      state.zoom = false;
      state.visible = false;
      state.collapsed = false;
      state.error = true;
      state.information = false;
      state.tab = 0;
      state.data = {};
      const query = new URLSearchParams(window.location.search);
      query.delete('egrid');
      query.delete('identdn');
      query.delete('number');
      window.history.pushState(null, null, '?' + query.toString());
    },
    hideExtract: (state) => {
      state.loading = false;
      state.zoom = false;
      state.visible = false;
      state.collapsed = false;
      state.error = false;
      state.egrid = null;
      state.identdn = null;
      state.number = null;
      state.information = false;
      state.tab = 0;
      state.data = {};
      const query = new URLSearchParams(window.location.search);
      query.delete('egrid');
      query.delete('identdn');
      query.delete('number');
      window.history.pushState(null, null, '?' + query.toString());
    },
    toggleCollapsed: (state) => {
      state.collapsed = !state.collapsed;
    },
    toggleHighlight: (state) => {
      state.highlight = !state.highlight;
    },
    toggleBackground: (state) => {
      state.background = !state.background;
    },
    toggleInformationPanel: (state) => {
      state.information = !state.information;
      if (state.information) {
        state.tab = 0;
      }
    },
    setInformationPanelTab: (state, action) => {
      state.tab = action.payload;
    }
  }
});

export const {
  loadExtract,
  showExtract,
  showError,
  hideExtract,
  toggleCollapsed,
  toggleHighlight,
  toggleBackground,
  toggleInformationPanel,
  setInformationPanelTab
} = extractSlice.actions;

export default extractSlice.reducer;
