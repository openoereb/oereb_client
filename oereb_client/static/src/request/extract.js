import {isString} from "lodash";

export const queryExtractById = function (serviceUrl, egrid, identdn, number, timeout, language) {
  const url = new URL(serviceUrl + 'extract/json/');
  if (isString(egrid)) {
    url.searchParams.append('EGRID', egrid);
  }
  else {
    url.searchParams.append('IDENTDN', identdn);
    url.searchParams.append('NUMBER', number);
  }
  url.searchParams.append('GEOMETRY', true);
  url.searchParams.append('_dc', new Date().getTime());
  if (language) {
    url.searchParams.append('LANG', language);
  }
  return new Promise((resolve, reject) => {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout * 1000);
    fetch(url.toString(), {signal: controller.signal})
      .then((response) => {
        clearTimeout(timer);
        if (response.ok) {
          resolve(response.json());
        }
        else {
          reject(new Error(response.text()));
        }
      })
      .catch((error) => {
        clearTimeout(timer);
        reject(error);
      });
  });
};

export const queryStaticExtractById = function (serviceUrl, egrid, identdn, number, timeout, language) {
  const url = new URL(serviceUrl + 'extract/pdf/');
  if (isString(egrid)) {
    url.searchParams.append('EGRID', egrid);
  }
  else {
    url.searchParams.append('IDENTDN', identdn);
    url.searchParams.append('NUMBER', number);
  }
  url.searchParams.append('_dc', new Date().getTime());
  if (language) {
    url.searchParams.append('LANG', language);
  }
  return new Promise((resolve, reject) => {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout * 1000);
    fetch(url.toString(), {signal: controller.signal})
      .then((response) => {
        clearTimeout(timer);
        if (response.ok) {
          resolve(response.arrayBuffer());
        }
        else {
          reject(new Error(response.text()));
        }
      })
      .catch((error) => {
        clearTimeout(timer);
        reject(error);
      });
  });
};

export const sanitizeTopicCode = function (theme) {
  let code = theme.Code.replace('.', '').replace('-', '').replace('_', '');
  if (isString(theme.SubCode)) {
    code += '_' + theme.SubCode.replace('.', '').replace('-', '').replace('_', '');
  }
  return code;
};

export const getParentTheme = function(theme, concernedThemes) {
  let result = null;
  for (let i = 0; i < concernedThemes.length; i++) {
    if (concernedThemes[i].Code === theme.Code) {
      result = concernedThemes[i];
      break;
    }
  }
  return result;
};

export const groupRestrictionsByTopic = function (restrictions, concernedThemes) {
  const restrictionsByTopic = {};
  restrictions.forEach((restriction) => {
    const code = sanitizeTopicCode(restriction.Theme);
    const lawstatus = restriction.Lawstatus.Code;
    if (!Reflect.apply(Object.prototype.hasOwnProperty, restrictionsByTopic, [code])) {
      let themeText = null;
      let subThemeText = null;
      if (isString(restriction.Theme.SubCode)) {
        subThemeText = restriction.Theme.Text;
        themeText = getParentTheme(restriction.Theme, concernedThemes).Text;
      }
      else {
        themeText = restriction.Theme.Text;
      }
      restrictionsByTopic[code] = {
        themeText: themeText,
        subThemeText: subThemeText,
        inForce: [],
        changeWithPreEffect: [],
        changeWithoutPreEffect: []
      };
    }
    restrictionsByTopic[code][lawstatus].push(restriction);
  });
  return restrictionsByTopic;
};
