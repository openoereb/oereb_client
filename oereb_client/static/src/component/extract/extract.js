import './extract.scss';

import React from 'react';
import { useSelector } from 'react-redux';
import OerebExtractLoading from './extract_loading';
import OerebExtractError from './extract_error';
import OerebExtractData from './extract_data';

function OerebExtract(props) {
    const extract = useSelector((state) => state.extract);

    function getContent(cfg) {
        if (cfg.loading) {
            return (
                <OerebExtractLoading />
            );
        }
        else if (cfg.error) {
            return (
                <OerebExtractError />
            );
        }
        else if (cfg.visible) {
            return (
                <OerebExtractData data={extract.data.GetExtractByIdResponse.extract} />
            );
        }
        else {
            return null;
        }
    }

    const content = getContent(extract);

    return (
        <div class="oereb-client-extract">
            {content}
        </div>
    );
}

export default OerebExtract;
