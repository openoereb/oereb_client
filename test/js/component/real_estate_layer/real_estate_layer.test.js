import {render} from '@testing-library/react';
import {Map} from "ol";
import VectorLayer from 'ol/layer/Vector';
import VectorSource from "ol/source/Vector";
import React from "react";
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";

import OerebRealEstateLayer from
  "../../../../oereb_client/static/src/component/real_estate_layer/real_estate_layer";
import {loadExtract, showExtract} from "../../../../oereb_client/static/src/reducer/extract";
import MainStore from "../../../../oereb_client/static/src/store/main";
import extract from "../../../../samples/extract.json";

describe('real estate layer component', () => {

  let component;

  beforeEach(() => {
    const map = new Map();
    const layer = new VectorLayer({
      source: new VectorSource()
    });
    map.addLayer(layer);
    act(() => {
      MainStore.dispatch(loadExtract({
        egrid: 'CH1234',
        zoom: false
      }));
    });
    act(() => {
      MainStore.dispatch(showExtract(extract));
    });
    component = render(
      <Provider store={MainStore}>
        <OerebRealEstateLayer realEstateLayer={layer} map={map} />
      </Provider>
    );
  });

  it('should render real estate layer element', () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

});