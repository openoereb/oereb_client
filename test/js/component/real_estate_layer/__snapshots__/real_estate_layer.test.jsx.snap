// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`real estate layer component > should render real estate layer element 1`] = `<DocumentFragment />`;
